### PR TITLE
refactor: runtime cleanup

### PR DESCRIPTION
**Pull request by 3Hren**
- Drop SFINAE magic in signal registering, because fuck you, that’s why.
- Added signal engine façade, which encapsulates signal handling
runtime and allows automatically join/cancel thread/cancellations.
- Fixed ignored logging backend parameter (always used «core» before).
- Replaced hardcoded return codes with defined ones.
- Minor renaming and includes elimination.

@antmat @karitra TAL.

As the result: `runtime.cpp` looks small and clean again.

### CODE REVIEW

File: src/signal.cpp [hunk 1]
```diff
@@ -271,4 +277,104 @@ handler_t::async_wait_simple(int signum, simple_callback_type handler) {
...
+        on(signum, terminate_handler_t());
+    }
+
+    // Wait until signaling termination.
+    std::unique_lock<std::mutex> lock(finalizer_mutex);
+    finalizer_cv.wait(lock, [&] { return finalize; });
+
+    // Unlock the mutex, as we don't need it anymore to prevent deadlock with several terminate
+    // calls.
+    lock.unlock();
```
Comment (karitra) [inline src/signal.cpp]: Is it default behaviour for `lock` on leaving the scoop?

File: include/cocaine/signal.hpp [hunk 2]
```diff
@@ -71,29 +80,14 @@ class handler_base_t {
 
     // Workaround for stupid gcc 4.6 which cannot choose overload
```
Comment (antmat) [inline include/cocaine/signal.hpp]: Drop the comment

File: src/signal.cpp [hunk 3]
```diff
@@ -271,4 +277,104 @@ handler_t::async_wait_simple(int signum, simple_callback_type handler) {
     });
 }
 
+namespace {
+
+namespace ph = std::placeholders;
+
+bool finalize = false;
```
Comment (antmat) [inline src/signal.cpp]: I think you can put it inside engine now

File: src/signal.cpp [hunk 4]
```diff
@@ -271,4 +277,104 @@ handler_t::async_wait_simple(int signum, simple_callback_type handler) {
...
+    std::unique_lock<std::mutex> lock(finalizer_mutex);
+    finalizer_cv.wait(lock, [&] { return finalize; });
+
+    // Unlock the mutex, as we don't need it anymore to prevent deadlock with several terminate
+    // calls.
+    lock.unlock();
+}
+
+auto
+engine_t::run() -> void {
```
Comment (antmat) [inline src/signal.cpp]: why not chaining?

**Pull request merged**